receiver/oktareceiver/logs.go

```diff
@@ -0,0 +1,236 @@
+	Do(req *http.Request) (*http.Response, error)
+	CloseIdleConnections()
+}
+
+// newOktaLogsReceiver returns a newly configured oktaLogsReceiver
+func newOktaLogsReceiver(cfg *Config, logger *zap.Logger, consumer consumer.Logs) (*oktaLogsReceiver, error) {
+	return &oktaLogsReceiver{
+		cfg:      *cfg,
+		client:   http.DefaultClient,
+		consumer: consumer,
+		logger:   logger,
+		wg:       &sync.WaitGroup{},
+	}, nil
+}
+
+func (r *oktaLogsReceiver) Start(_ context.Context, _ component.Host) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	r.cancel = cancel
+	r.wg.Add(1)
+	go r.startPolling(ctx)
+	return nil
+}
+
+func (r *oktaLogsReceiver) startPolling(ctx context.Context) {
+	defer r.wg.Done()
+	t := time.NewTicker(r.cfg.PollInterval)
+
+	err := r.poll(ctx)
+	if err != nil {
+		r.logger.Error("there was an error during the first poll", zap.Error(err))
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-t.C:
+			err := r.poll(ctx)
+			if err != nil {
+				r.logger.Error("there was an error during the poll", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (r *oktaLogsReceiver) poll(ctx context.Context) error {
+	logEvents := r.getLogs(ctx)
+	observedTime := pcommon.NewTimestampFromTime(time.Now())
+	logs := r.processLogEvents(observedTime, logEvents)
+	if logs.LogRecordCount() > 0 {
+		if err := r.consumer.ConsumeLogs(ctx, logs); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *oktaLogsReceiver) getLogs(ctx context.Context) []*okta.LogEvent {
+	var logs []*okta.LogEvent
+	var reqURL string
+	pollTime := time.Now().UTC()
+
+	// get logs until there isn't any overflow OR we get logs published after initial pollTime
+	for {
+		if r.nextURL == "" {
+			reqURL = "https://" + r.cfg.Domain + "/api/v1/logs"
+		} else {
+			reqURL = r.nextURL
+		}
+
+		req, err := http.NewRequestWithContext(ctx, "GET", reqURL, nil)
+		if err != nil {
+			r.logger.Error("error creating okta api request", zap.Error(err))
+			break
+		}
+
+		// add query params to the first polling request
+		if r.nextURL == "" {
+			query := req.URL.Query()
+			query.Add("since", pollTime.Add(-r.cfg.PollInterval).Format(OktaTimeFormat))
+			query.Add("limit", strconv.Itoa(oktaMaxLimit))
+			req.URL.RawQuery = query.Encode()
+		}
+
+		req.Header.Add("Authorization", "SSWS "+string(r.cfg.APIToken))
+
+		res, err := r.client.Do(req)
+		if err != nil {
+			r.logger.Error("error performing okta api request", zap.Error(err))
+			break
+		}
+
+		defer res.Body.Close()
+
+		if res.StatusCode != http.StatusOK {
+			r.logger.Error("okta logs endpoint returned non-200 statuscode: " + res.Status)
+			break
+		}
+
+		r.setNextLink(res)
+
+		body, err := io.ReadAll(res.Body)
+		if err != nil {
+			r.logger.Error("error reading response body", zap.Error(err))
+			break
+		}
+
+		var curLogs []*okta.LogEvent
+		err = json.Unmarshal(body, &curLogs)
+		if err != nil {
+			r.logger.Error("unable to unmarshal log events", zap.Error(err))
+			break
+		}
```

Comment:
We don't necessarily have to do this here, but I was thinking about this and if we wanted to preserve the original JSON exactly, we could by unmarshalling into [json.RawMessage](https://pkg.go.dev/encoding/json#RawMessage) and later decoding that into okta.LogEvent when converting to a log record. This would let you keep the original representation of the log.

Reply:
What would be the motivation to do that instead of what I wrote? So we would do the unmarshalling later in the process and an error wouldn't hold up the request cycle?

Reply:
The motivation is just to preserve the original log byte-for-byte. I don't think we need to do it, to be clear. Just a thought I had.